### DISCUSSION TITLE
smol in sync / non-blocking context

### DISCUSSION THREAD

**Discussion by Roms1383** [smol-rs/smol#297] [Q&A] [answered]
Hello, first and foremost thanks for this library.
This morning I made an experiment that is an unusual context and hits a strange behaviour that you might be able to help me understand.

I'm currently developing a native plugin for game modding with a Rust crate (namely [red4ext-rs](https://github.com/jac3km4/red4ext-rs)).
On game startup on Windows I load a bunch of audio files and subtitles from a YAML.
So far I do it without an async runtime, simply with standard methods.

Foreseeing that in the future multiple mods could be loading many YAML files with dozens of audio file paths and subtitles, I was considering adding `smol::Executor` and `spawn`ing a `Task` that I would directly `detach` to avoid blocking the thread, letting it load silently in the background instead of slowing down game startup (at least that's the assumption I have, let me know if it's incorrect).

My experiment is roughly like this:
```rs
lazy_static! {
    static ref EXECUTOR: Executor<'static> = Executor::new();
}

pub fn setup() -> anyhow::Result<()> {
    EXECUTOR.spawn(async {
        match banks::setup().await { // where banks::setup() is an async fn, the one that loads everything
            // both of these lines simply log to a file
            Ok(_) => { red4ext_rs::info!("banks loaded successfully!"); }
            Err(e) => { red4ext_rs::error!("failed to load banks ({e})"); }
        }
    }).detach();

    // some more setup ...
    Ok(())
}
```

The result is that there's no crash, but no success or error message is written to the logs.
Furthermore in-game it's like my audio banks have never been loaded at all (while it usually does nicely).

Do you know by any chance what I could possibly be missing?
Do I need further async setup like a channel to communicate when `banks::setup()` is successful / has failed ?

Thanks in advance 😃

**Comment by notgull** [comment, marked as answer]
You need to call [`run`](https://docs.rs/async-executor/latest/async_executor/struct.Executor.html#method.run) to drive the future. Executors on their own don't run, you need to drive them. For example:

```
std::thread::spawn(|| {
    smol::block_on(EXECUTOR.run(std::future::pending::<()>()));
});
```

**Reply by Roms1383** [reply]
Indeed, utterly makes sense, thanks.
